### PR TITLE
Effects type classes

### LINKED ISSUE #253: Bracket
## Description

Add `Bracket` type class to the Effects module. See its implementation [in Arrow](https://github.com/arrow-kt/arrow/blob/master/modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/typeclasses/Bracket.kt). See its [type class laws](https://github.com/arrow-kt/arrow/blob/b5746f09d505a53b77f5aa6b247b01536ab45045/modules/core/arrow-test/src/main/kotlin/arrow/test/laws/BracketLaws.kt).

## Modules

- `BowEffects`
- `BowEffectsLaws`

### LINKED ISSUE #254: MonadDefer
## Description

Update `MonadDefer` to the definition of the [type class in Arrow](https://github.com/arrow-kt/arrow/blob/master/modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/typeclasses/MonadDefer.kt). See its [type class laws](https://github.com/arrow-kt/arrow/blob/b5746f09d505a53b77f5aa6b247b01536ab45045/modules/core/arrow-test/src/main/kotlin/arrow/test/laws/MonadDeferLaws.kt).

## Modules

- `BowEffects`.
- `BowEffectsLaws`.

## Breaking changes

Changes in the current implementation of `MonadDefer` may break existing instances in the modules for `BowRx` and `BowBrightFutures`, and in the `IO` instances.

### LINKED ISSUE #256: Async
## Description

Update to the definition of the [type class in Arrow](https://github.com/arrow-kt/arrow/blob/master/modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/typeclasses/Async.kt). See its [type class laws](https://github.com/arrow-kt/arrow/blob/b5746f09d505a53b77f5aa6b247b01536ab45045/modules/core/arrow-test/src/main/kotlin/arrow/test/laws/AsyncLaws.kt).

## Modules

- `BowEffects`.
- `BowEffectsLaws`.

## Breaking changes

Existing instances may break if `Async` is changed; particularly, `IO` and instances in `BowRx` and `BowBrightFutures`.

### LINKED ISSUE #257: Effect
## Description

Review `Effect` type class to make sure it is aligned with the [definition in Arrow](https://github.com/arrow-kt/arrow/blob/master/modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/typeclasses/Effect.kt).

## Modules

- `BowEffects`.
- `BowEffectsLaws`.

### LINKED ISSUE #260: ConcurrentEffect
## Description

Review `ConcurrentEffect` to check everything is correct after implementing the rest of the type classes. See its [definition in Arrow](https://github.com/arrow-kt/arrow/blob/master/modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/typeclasses/ConcurrentEffect.kt).

## Modules

- `BowEffects`.

### LINKED ISSUE #262: UnsafeRun
## Description

Add `UnsafeRun` type class as per its [definition in Arrow](https://github.com/arrow-kt/arrow/blob/master/modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/typeclasses/UnsafeRun.kt). This is defined in terms of `unsafe`, which is an object wrapper using coroutines. This needs to be adjusted to the Swift platform using the corresponding equivalence.

## Modules

- `BowEffects`.

### LINKED ISSUE #266: Fiber
## Description

Add the notion of `Fiber` as per its [definition in Arrow](https://github.com/arrow-kt/arrow/blob/master/modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/typeclasses/Fiber.kt).

## Modules

- `BowEffects`.

### LINKED ISSUE #287: KindConnection
## Description

Add `KindConnection` as per its [definition in Arrow](https://github.com/arrow-kt/arrow/blob/master/modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/KindConnection.kt).

## Modules

- `BowEffects`.

### PR DESCRIPTION
**Pull request by truizlop**
## Related issues

- Closes #253 
- Closes #254 
- Closes #256 
- Closes #257 
- Addresses partially #258 
- Closes #287 
- Closes #266 
- Closes #260 
- Closes #262 

## Goal

Include type classes to work with effects.

## Implementation details

The following type classes have been added or reworked:

- Bracket
- MonadDefer
- Async
- Effect
- Concurrent
- ConcurrentEffect
- UnsafeRun

In order to implement these type classes, the following types were needed:

- Fiber
- KindConnection

## Testing details

Laws for the following type classes have been added or reworked:

- Async
- Bracket
- MonadDefer

Laws for Concurrent are missing in this PR since the need other existing types like Promise or MVar that will be worked in another PR.

### CODE REVIEW

**Comment by truizlop** [comment]
Reference in Arrow:
- Type classes: https://github.com/arrow-kt/arrow/tree/master/modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/typeclasses
- Laws: https://github.com/arrow-kt/arrow/tree/master/modules/core/arrow-test/src/main/kotlin/arrow/test/laws

File: Tests/BowEffectsLaws/AsyncLaws.swift [hunk 1]
```diff
@@ -3,24 +3,78 @@ import SwiftCheck
...
+public class AsyncLaws<F: Async & EquatableK> where F.E: Arbitrary {
     
-    static func check() {
+    public static func check() {
         success()
         error()
+        continueOnJumpsThreads()
+        asyncConstructor()
+        continueOnComprehension()
+        asyncCanBeDerivedFromAsyncF()
```
Comment (miguelangel-dev) [inline Tests/BowEffectsLaws/AsyncLaws.swift]: missing `bracketReleaseIscalledOnCompletedOrError`, is it possible to add it?

Comment (truizlop) [reply Tests/BowEffectsLaws/AsyncLaws.swift]: Oh I see. That test depends on `Promise`, which is still not created. I will add the test once Promise is added.

Comment (miguelangel-dev) [reply Tests/BowEffectsLaws/AsyncLaws.swift]: Great!

**Pull request merged**